File: clients/python/aether_client/withdraw.py
"""Taking back unspent prepaid balance.

The seller pays it back on chain to the signing account itself; each
withdrawal_id pays out at most once, so retry with the same one.
"""

import json
import os
import re
import time
import urllib.parse
from dataclasses import dataclass
from typing import Optional

from .amount import parse_amount, parse_uaeth
from .client import AetherClient
from .directory import MANIFEST_PATH
from .keys import Key
from .paywall import PaymentError, _http, prepaid_payment_header

_CODES = {
    "withdrawals_unavailable": "WITHDRAWALS_UNAVAILABLE",
    "insufficient_balance": "INSUFFICIENT_PREPAID_BALANCE",
    "below_minimum_withdrawal": "INSUFFICIENT_PREPAID_BALANCE",
    "withdrawal_id_reused": "IDEMPOTENCY_CONFLICT",
    "payout_failed": "WITHDRAWAL_FAILED",  # nothing paid; the balance is intact
    "payout_unavailable": "WITHDRAWAL_FAILED",
}


@dataclass
class WithdrawResult:
    status: str  # pending (sent, not in a block yet) | confirmed | reserved (not sent yet: call again, same id)
    withdrawal_id: str
    amount_uaeth: int
    tx_hash: Optional[str] = None
    balance_uaeth: Optional[int] = None  # left with the seller
    message: Optional[str] = None


def withdraw_prepaid(client: AetherClient, key: Key, service: str, amount: str = "all",
                     withdrawal_id: Optional[str] = None) -> WithdrawResult:
    """Withdraws unspent prepaid balance from the service at `service` (any URL on it).
    amount is "all" or carries its unit ("0.5 AETH")."""
    u = urllib.parse.urlsplit(service)
    if u.scheme not in ("http", "https") or not u.netloc:
        raise PaymentError("INVALID_ARGUMENT", f"service {service} must be an http(s) URL")
    origin = f"{u.scheme}://{u.netloc}"
    want = "all" if not amount or amount.strip().lower() == "all" else str(parse_amount(amount))
    withdrawal_id = withdrawal_id or os.urandom(16).hex()

    m = _http("GET", origin + MANIFEST_PATH, b"", {})
    if m.status != 200:
        raise PaymentError("PAYMENT_UNSUPPORTED", f"{origin}{MANIFEST_PATH} answered HTTP {m.status}: not an Aether paid service")
    manifest = json.loads(m.body)
    if manifest.get("network") != client.chain_id:
        raise PaymentError("PAYMENT_UNSUPPORTED", f"the service is on network {manifest.get('network')}, not {client.chain_id}")
    path = manifest.get("withdrawPath")
    if not path:
        raise PaymentError("WITHDRAWALS_UNAVAILABLE", "this service doesn't offer withdrawals: its operator holds the balance")
    # A path, never something that would change the host when appended ("@evil.example/").
    if not isinstance(path, str) or not path.startswith("/") or path.startswith("//"):
        raise PaymentError("PAYMENT_UNSUPPORTED", "the manifest's withdrawPath is not a path on this service")

    body = json.dumps({"amount": want}, separators=(",", ":")).encode()
    header = prepaid_payment_header(key, network=client.chain_id, pay_to=manifest["payTo"], host=u.netloc, method="POST",
                                    path=path, body=body, max_price=0, timestamp=int(time.time()), request_id=withdrawal_id)
    resp = _http("POST", origin + path, body, {"Content-Type": "application/json", "X-PAYMENT": header})
    try:
        r = json.loads(resp.body)
    except ValueError:
        raise PaymentError("HTTP_ERROR", f"the service answered HTTP {resp.status} without a withdrawal result")
    if r.get("error"):
        raise PaymentError(_CODES.get(r["error"], "WITHDRAWAL_REJECTED"),
                           f"the service refused the withdrawal ({r['error']}): {r.get('message', '')}")
    bal = r.get("balance")
    return WithdrawResult(r.get("status", "pending"), withdrawal_id, parse_uaeth(r["amount"]) if r.get("amount") else 0,
                          r.get("txHash"), int(bal) if isinstance(bal, str) and re.fullmatch(r"[0-9]+", bal) else None, r.get("message"))

File: clients/ts/src/withdraw.ts
import { bytesToHex } from "@noble/hashes/utils.js";
import { AetherClient } from "./client.js";
import { Key } from "./keys.js";
import { parseAmount, parseUaeth } from "./amount.js";
import { MANIFEST_PATH, type Manifest } from "./directory.js";
import { PaymentError, prepaidPaymentHeader } from "./paywall.js";

// Taking back unspent prepaid balance. The seller pays it back on chain
// to the signing account itself; each withdrawalId pays out at most
// once, so retry with the same one.

export interface WithdrawOptions {
  /** "all" (default) or an amount with its unit ("0.5 AETH"). */
  amount?: string;
  /** Reuse it to retry: the seller pays each ID out once. Default: random. */
  withdrawalId?: string;
}

export interface WithdrawResult {
  /** pending: sent, not in a block yet; confirmed; reserved: set aside, not sent yet -- call again with the same withdrawalId. */
  status: "pending" | "confirmed" | "reserved";
  withdrawalId: string;
  amountUaeth: bigint;
  txHash?: string;
  /** Left with the seller. */
  balanceUaeth?: bigint;
  message?: string;
}

interface WithdrawalResponse {
  withdrawalId?: string;
  amount?: string;
  status?: WithdrawResult["status"];
  txHash?: string;
  balance?: string;
  error?: string;
  message?: string;
}

const CODES: Record<string, string> = {
  withdrawals_unavailable: "WITHDRAWALS_UNAVAILABLE",
  insufficient_balance: "INSUFFICIENT_PREPAID_BALANCE",
  below_minimum_withdrawal: "INSUFFICIENT_PREPAID_BALANCE",
  withdrawal_id_reused: "IDEMPOTENCY_CONFLICT",
  payout_failed: "WITHDRAWAL_FAILED", // nothing paid; the balance is intact
  payout_unavailable: "WITHDRAWAL_FAILED",
};

/** Withdraws unspent prepaid balance from the service at `service` (any URL on it). */
export async function withdrawPrepaid(client: AetherClient, key: Key, service: string, opts: WithdrawOptions = {}): Promise<WithdrawResult> {
  const u = new URL(service);
  if (u.protocol !== "http:" && u.protocol !== "https:") throw new PaymentError("INVALID_ARGUMENT", `service ${service} must be an http(s) URL`);
  const amount = !opts.amount || opts.amount.trim().toLowerCase() === "all" ? "all" : parseAmount(opts.amount).toString();
  const withdrawalId = opts.withdrawalId ?? bytesToHex(crypto.getRandomValues(new Uint8Array(16)));

  const mresp = await fetch(u.origin + MANIFEST_PATH, { redirect: "manual" });
  if (mresp.status !== 200) throw new PaymentError("PAYMENT_UNSUPPORTED", `${u.origin}${MANIFEST_PATH} answered HTTP ${mresp.status}: not an Aether paid service`);
  const m = (await mresp.json()) as Manifest;
  if (m.network !== client.chainId) throw new PaymentError("PAYMENT_UNSUPPORTED", `the service is on network ${m.network}, not ${client.chainId}`);
  if (!m.withdrawPath) throw new PaymentError("WITHDRAWALS_UNAVAILABLE", "this service doesn't offer withdrawals: its operator holds the balance");
  // A path, never something that would change the host when appended ("@evil.example/").
  if (!/^\/(?!\/)/.test(m.withdrawPath)) throw new PaymentError("PAYMENT_UNSUPPORTED", "the manifest's withdrawPath is not a path on this service");

  const body = new TextEncoder().encode(JSON.stringify({ amount }));
  const header = prepaidPaymentHeader(key, {
    network: client.chainId, payTo: m.payTo, host: u.host, method: "POST", path: m.withdrawPath,
    body, maxPrice: 0n, timestamp: Math.floor(Date.now() / 1000), requestId: withdrawalId,
  });
  const resp = await fetch(u.origin + m.withdrawPath, {
    method: "POST", redirect: "manual", headers: { "Content-Type": "application/json", "X-PAYMENT": header },
    body: body as unknown as BodyInit,
  });
  let r: WithdrawalResponse;
  try {
    r = (await resp.json()) as WithdrawalResponse;
  } catch {
    throw new PaymentError("HTTP_ERROR", `the service answered HTTP ${resp.status} without a withdrawal result`);
  }
  if (r.error) throw new PaymentError(CODES[r.error] ?? "WITHDRAWAL_REJECTED", `the service refused the withdrawal (${r.error}): ${r.message ?? ""}`);
  return {
    status: r.status ?? "pending", withdrawalId, amountUaeth: r.amount ? parseUaeth(r.amount) : 0n, txHash: r.txHash,
    balanceUaeth: r.balance !== undefined && /^\d+$/.test(r.balance) ? BigInt(r.balance) : undefined, message: r.message,
  };
}
